### PR TITLE
Add hooks for dynamic completion

### PR DESCRIPTION
**Pull request by dbanck**
This PR introduces a new `hooks` package, which we can use to implement hooks for dynamic completion. Hooks will be implemented as part of https://github.com/hashicorp/vscode-terraform/issues/672.

We also introduce a new `completionItem/resolve` handler, which editors can use to update a completion item with more detailed data. The data is obtained via the execution of a resolve hook.

---

* Closes https://github.com/hashicorp/terraform-ls/issues/868
* Requires https://github.com/hashicorp/hcl-lang/pull/123

### CODE REVIEW

File: internal/decoder/decoder.go [hunk 1]
```diff
@@ -79,11 +73,13 @@ func varsPathContext(mod *state.Module) (*decoder.PathContext, error) {
...
+func DecoderContext(ctx context.Context) decoder.DecoderContext {
 	dCtx := decoder.DecoderContext{
-		UtmSource:     utm.UtmSource,
-		UtmMedium:     utm.UtmMedium(ctx),
-		UseUtmContent: true,
+		UtmSource:              utm.UtmSource,
+		UtmMedium:              utm.UtmMedium(ctx),
+		UseUtmContent:          true,
+		CompletionHooks:        make(decoder.CompletionFuncMap),
+		CompletionResolveHooks: make(decoder.CompletionResolveFuncMap),
```
Comment (radeksimko) [inline internal/decoder/decoder.go:82]: As mentioned in the other PR, ideally this shouldn't be necessary.

File: internal/hooks/hooks.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Package hooks enables the implementation of hooks for dynamic
+// autocompletion. Hooks should be added to this package and
+// registered in AppendCompletionHooks in completion_hooks.go.
```
Comment (radeksimko) [inline internal/hooks/hooks.go:3]:
```suggestion
// registered via AppendCompletionHooks in completion_hooks.go.
```

File: internal/hooks/hooks.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+//
+// A hook must have the following signature:
+//  func (h *Hooks) Name(ctx context.Context, value cty.Value) ([]decoder.Candidate, error)
+// It receives the current value of the attribute and must return
+// a list of completion candidates.
```
Comment (radeksimko) [inline internal/hooks/hooks.go:8]: I would argue this is something that a compile-time error would explain and it would likely be even more accurate. As much as I appreciate the commenting efforts, the risk here is that we end up changing the interface upstream but won't remember to update this comment. Compiler will however always bring it up if types/interfaces don't match.

We could just refer to `decoder.CompletionFunc` here and document that one upstream instead.

File: internal/hooks/hooks.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//
+// All hooks have access to path, filename and pos via context:
+//  path, ok := decoder.PathFromContext(ctx)
+//  filename, ok := decoder.FilenameFromContext(ctx)
+//  pos, ok := decoder.PosFromContext(ctx)
```
Comment (radeksimko) [inline internal/hooks/hooks.go:13]: This is very useful context, but I think that should be part of docs for `CompletionFunc` upstream, where we can more easily keep it up to date.

File: internal/hooks/module_source_local.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package hooks
+
+import (
+	"context"
+
+	"github.com/hashicorp/hcl-lang/decoder"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func (h *Hooks) LocalModuleSources(ctx context.Context, value cty.Value) ([]decoder.Candidate, error) {
```
Comment (radeksimko) [inline internal/hooks/module_source_local.go:10]: I'm assuming this would be eventually part of a separate PR? 👀

File: internal/langserver/handlers/completion_resolve.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	"context"
+
+	"github.com/hashicorp/hcl-lang/decoder"
+	"github.com/hashicorp/hcl-lang/lang"
+	ilsp "github.com/hashicorp/terraform-ls/internal/lsp"
+	"github.com/hashicorp/terraform-ls/internal/mdplain"
+	lsp "github.com/hashicorp/terraform-ls/internal/protocol"
+)
+
+type CompletionItem struct {
```
Comment (radeksimko) [inline internal/langserver/handlers/completion_resolve.go:13]: Nitpick: We could move this under the `protocol` package as we already have some extensions of the original structs there for telemetry or experimental functions:
https://github.com/hashicorp/terraform-ls/tree/main/internal/protocol

File: internal/lsp/completion.go [hunk 7]
```diff
@@ -70,7 +70,12 @@ func toCompletionItem(candidate lang.Candidate, caps lsp.CompletionClientCapabil
 		Documentation:       doc,
 		TextEdit:            textEdit(candidate.TextEdit, snippetSupport),
 		Command:             cmd,
-		AdditionalTextEdits: textEdits(candidate.AdditionalTextEdits, snippetSupport),
+		AdditionalTextEdits: TextEdits(candidate.AdditionalTextEdits, snippetSupport),
+		// TODO set deprecated via `tags` if supported.
```
Comment (radeksimko) [inline internal/lsp/completion.go:74]: Thanks for adding the TODO, is it worth moving it below, where we use the legacy `DeprecatedSupport` bool field?

**Review by radeksimko — APPROVED** [review]
LGTM, aside from a few in-line comments, although you may want to look at the upstream `hcl-lang` PR first and then rebase here once merged.

File: internal/hooks/module_source_local.go (hunk 5 above)
Comment (dbanck) [reply internal/hooks/module_source_local.go:10]: I added this as a placeholder to showcase the registration of hooks in `AppendCompletionHooks`. Without any hooks we wouldn't be able to create `hooks.Hooks{}`.

But I can remove it if you want? And defer everything to the next PR

File: internal/lsp/completion.go (hunk 7 above)
Comment (dbanck) [reply internal/lsp/completion.go:74]: I missed below. We're already using tags

File: internal/protocol/completion.go [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package protocol
+
+import "github.com/hashicorp/hcl-lang/lang"
+
+type CompletionItemR struct {
```
Comment (radeksimko) [inline internal/protocol/completion.go:5]:
```suggestion
type CompletionItemWithResolveHook struct {
```
Just for some extra clarity
